### ISSUE TITLE
Base `include module` subdirectory on module, not file, name

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#15420]
While reviewing the recent change in PR #15279, it felt to me like it would be both convenient and natural to:

* permit the `include module` statement to appear in files whose modules don't match their names (e.g., `M-1.2.chpl` declares a module named `M`).
* look for the submodules to include in a directory based on the parent module name rather than the filename in such cases

Specifically, it seemed annoying / unfortunate to me that every test of `include module` had to have its own directory of submodule files rather than creating a smaller number of such subdirectories for a number of similar tests to share.  And in my own work, I often make file clones of given source files when I'm experimenting with them, and would find it a hassle to have to create symbolic links or copies of the submodule directories each time I did.

**Comment by lydia-duncan** [comment]
I think I agree with this.  I have some vague concerns about not knowing what parent files are impacted when changing a submodule due to not being able to tell based on their file names, but I think that's true of any library that is used or imported.  Perhaps @BryantLam can think of a case I'm not thinking of, I think he has a better grasp on that category of concerns than I do

**Comment by BryantLam** [comment]
> * permit the `include module` statement to appear in files whose modules don't match their names (e.g., `M-1.2.chpl` declares a module named `M`).

I don't have strong concerns with this. I find it undesirable that the named module `M` is allowed to diverge with a different filename like `M-1.2.chpl` *but* the impact is limited by the constraint that the filename can't be used directly without a symlink to `M.chpl` so there's apparent traceability (instead of implicit traceability where I have to go digging through files to figure out what is used where).

> * look for the submodules to include in a directory based on the declared module name rather than the filename in such cases

Like this?
```
.
├── M -> M-1.2
├── M-1.2
│   └── Msub.chpl
├── M.chpl -> M-1.2.chpl
└── M-1.2.chpl # module M includes module Msub
```
I would assume the implementation would allow this usage. If not, what does it currently do?

> Specifically, it seemed annoying / unfortunate to me that every test of include module had to have its own directory of submodule files rather than creating a smaller number of such subdirectories for a number of similar tests to share.

It probably means those tests shouldn't be in a submodule.. but the submodule feature is also being tested, which is a bit of a corner case compared to real-world libraries that can refactor and pull said tests to a top-level package `Test` module.

> And in my own work, I often make file clones of given source files when I'm experimenting with them, and would find it a hassle to have to create symbolic links or copies of the submodule directories each time I did.

This statement contradicts what I read from the second bullet, so I'm probably misunderstanding something.

**Comment by lydia-duncan** [comment]
I should also point out that `M-1.2` is not a valid module name anyways, so the file-scoped module would not be able to have submodules otherwise

**Comment by bradcray** [comment]
> This statement contradicts what I read from the second bullet, so I'm probably misunderstanding something.

I've added a concrete example to the OP to try and clarify.

**Comment by BryantLam** [comment]
I'm okay with this. One of my concerns is knowing how an executable was built from the codebase, but that should be somewhat alleviated by `<exe> --about`.

**Comment by bradcray** [comment]
> One of my concerns is knowing how an executable was built from the codebase

Is there something about this issue's feature request that makes that any harder than it currently is?

**Comment by BryantLam** [comment]
I don't think so for this specific feature request. General software engineering moreso.
